src/Facet/Driver.hs

```diff
@@ -90,46 +93,65 @@ kernel = Module kernelName [] [] $ Scope mempty
 reloadModules :: (Has (Error (Notice.Notice (Doc Style)) :+: Output :+: State Options :+: State Target :+: Write (Notice.Notice (Doc Style))) sig m, MonadIO m) => m ()
 reloadModules = do
   searchPaths <- uses searchPaths_ toList
+  modules_ .= singleton Nothing kernel
   modules <- targets_ ~> \ targets -> do
     -- FIXME: remove stale modules
+    -- FIXME: only reload changed modules
     -- FIXME: failed module header parses shouldn’t invalidate everything.
     targetHeads <- traverse (loadModuleHeader searchPaths . Right) (toList targets)
-    rethrowGraphErrors [] $ loadOrder (fmap toNode . loadModuleHeader searchPaths . Right) (map toNode targetHeads)
+    rethrowGraphErrors [] $ loadOrder (fmap headerNode . loadModuleHeader searchPaths . Right) (map headerNode targetHeads)
   let nModules = length modules
-  results <- evalFresh 1 $ for modules $ \ (name, path, src, imports) -> do
-    i <- fresh
-    outputDocLn $ annotate Progress (brackets (ratio i nModules)) <+> nest 2 (group (fillSep [ pretty "Loading", prettyMName name ]))
-
-    -- FIXME: skip gracefully (maybe print a message) if any of its imports are unavailable due to earlier errors
-    (Just <$> loadModule name path src imports) `catchError` \ err -> Nothing <$ outputDocLn (prettyNotice err)
+  results <- for (zip [1..] modules) $ \ (i, h@(ModuleHeader name src _)) -> do
+    graph <- use modules_
+    let loaded = traverse (\ name -> graph^.at name >>= snd) h
+    case loaded of
+      Just loaded -> (Just <$> do
+        outputDocLn $ annotate Progress (brackets (ratio (i :: Int) nModules)) <+> nest 2 (group (fillSep [ pretty "Loading", prettyMName name ]))
+        storeModule name (path src) =<< loadModule graph loaded)
+        `catchError` \ err -> Nothing <$ outputDocLn (prettyNotice err)
+      Nothing -> do
+        outputDocLn $ annotate Progress (brackets (ratio i nModules)) <+> nest 2 (group (fillSep [ pretty "Skipping", prettyMName name ]))
+        pure Nothing
```

Comment:
I’m thinking of extracting this loop for use in e.g. `Facet.Run`, but I think I’d also want to extract the logs here into something that carries some data about log levels so that we can filter it appropriately. Either that, or do it in some iterative manner that allows us to log only in the repl.